llvm/lib/ObjCopy/ELF/ELFObject.cpp

```diff
@@ -324,6 +324,14 @@ Expected<IHexRecord> IHexRecord::parse(StringRef Line) {
   return Rec;
 }
 
+static uint64_t sectionLMA(const SectionBase *Sec) {
```

Comment:
I'm curious about why this helper function is needed for SREC, if it didn't already exist for IHEX. The two seem similar to me: surely for both formats you want to calculate the memory addresses of segments the same way.

Does this patch agree with the way IHEX does it? If not, which do you think is right?

Reply:
In fact, now I look closer, the way IHEX does it is in the `sectionPhysicalAddr` function just below this new one. What's the reasoning for needing to do it differently?

Reply:
I see different results from gnu objcopy when using sectionPhysicalAddr. 

example using `ihex-elf-segments.yaml`:

GNU:
```
S00400007883
S214000000000102030405060708090A0B0C0D0E0F73
S20900001010111213148C
S20F00FFF83031323334353637383940AC
S20801010040414243EF
S20F10FFF850515253545556575859603C
S20720FFF8000000E1
S804100000EB

address data 

000000 000102030405060708090A0B0C0D0E0F
000010 1011121314
00FFF8 3031323334353637383940
010100 40414243
10FFF8 5051525354555657585960
20FFF8 000000
```

LLVM(sectionPhysicalAddr):
```
S00400007883
S214100000000102030405060708090A0B0C0D0E0F63
S20910001010111213147C
S20F10001830313233343536373839407B
S20810002840414243B9
S20F100030505152535455565758596003
S20720FFF8000000E1
S804100000EB

address data 

100000 000102030405060708090A0B0C0D0E0F
100010 1011121314
100018 3031323334353637383940
100028 40414243
100030 5051525354555657585960
20FFF8 000000
```

looking at the LMA values in objdump -h output, I think that the former is correct
```
objdump -h 1.o 

1.o:     file format elf64-x86-64

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .text         00000015  0000000000000000  0000000000000000  00000078  2**3
  1 .data1        0000000b  000000000000fff8  000000000000fff8  00000090  2**3
  2 .data2        00000004  0000000000010100  0000000000010100  000000a0  2**3
  3 .data3        0000000b  000000000010fff8  000000000010fff8  000000a8  2**3
```

Reply:
OK, so in this example ELF file, there are _three_ address layouts. (`llvm-readelf` or GNU `readelf` are better for showing this than `objdump`.)

Going by the addresses in the section header table, this image is highly non-contiguous, with `.text` going at 0, some data around 0x10000, some more data just below 0x200000 and more still just below 0x300000. Each section is only a few bytes long, and the gaps between them are many Kb or even Mb.

But then those sections are packed into a single `PT_LOAD` segment much more tightly, so that the only gaps between them are a few bytes for alignment. That segment has two different addresses: a virtual address (`p_vaddr`) of 0xF00000000, and a physical address (`p_paddr`) of 0x100000.

GNU objcopy, in both `ihex` and `srec` output mode, is using the layout from the section addresses, so that it emits records with addresses near 0, near 0x10000, near 0x200000 and near 0x300000.

`llvm-objcopy -O ihex` is using the layout from the segment's physical address, so that everything it emits is tightly packed near 0x100000.

`llvm-objcopy -O srec` with your patch is using the section addresses again, so that LLVM's two hex formats produce different addresses.

(And nothing is using the _virtual_ address of the segment, which is just as well, because it's greater than 2^32, and wouldn't fit into these 32-bit hex file formats in any case.)

I can see that you might want to be able to generate hex output based on the section addresses. But I think that the choice of addressing scheme should not depend on which hex format you're outputting. The two hex formats are used for similar purposes; there's no reason why one should be right for `srec` and the other one for `ihex`.

So I'd prefer that we use the same section-address function for both hex formats, and separately, introduce an option to control how that function decides on addresses.

Reply:
Thank you for the explanation.

Updated this patch to use the same section address function as ihex and updated a test to reflect this.